OpenEdXMobile/src/main/java/org/edx/mobile/view/dialog/FullscreenLoaderDialogFragment.kt

```diff
@@ -78,28 +84,47 @@ class FullscreenLoaderDialogFragment : DialogFragment() {
         outState.putLong(LOADER_START_TIME, loaderStartTime)
     }
 
+    fun setData(iapData: IAPFlowData) {
+        this.iapFlowData = iapData
+    }
+
     private fun intiViews() {
         binding.materialTextView.setText(getTitle(), TextView.BufferType.SPANNABLE)
     }
 
     private fun initObservers() {
-        iapViewModel.errorMessage.observe(viewLifecycleOwner, NonNullObserver { errorMsg ->
-            errorMsg.throwable as InAppPurchasesException
-            iapDialog.handleIAPException(
-                fragment = this@FullscreenLoaderDialogFragment,
-                errorMessage = errorMsg,
-                retryListener = { _, _ ->
-                    if (errorMsg.requestType == ErrorMessage.EXECUTE_ORDER_CODE) {
-                        iapViewModel.executeOrder(requireActivity())
-                    } else {
-                        iapViewModel.refreshCourseData(true)
-                    }
-                },
-                cancelListener = { _, _ ->
-                    resetPurchase()
-                })
-            iapViewModel.errorMessageShown()
+        iapViewModel.refreshCourseData.observe(viewLifecycleOwner, NonNullObserver {
+            closeLoader()
         })
+        iapViewModel.errorMessage.observe(viewLifecycleOwner, NonNullObserver { errorMessage ->
+            if (errorMessage.isPostUpgradeErrorType()) {
+                errorMessage.throwable as InAppPurchasesException
+                iapDialog.handleIAPException(
+                    fragment = this@FullscreenLoaderDialogFragment,
+                    errorMessage = errorMessage,
+                    retryListener = { _, _ ->
+                        if (errorMessage.requestType == ErrorMessage.EXECUTE_ORDER_CODE) {
+                            iapViewModel.executeOrder(iapFlowData)
+                        } else {
+                            closeLoader()
+                        }
+                    },
+                    cancelListener = { _, _ ->
+                        dismiss()
+                    })
+            }
+        })
+    }
+
+    private fun closeLoader() {
+        Timer("", false).schedule(getRemainingVisibleTime()) {
+            EventBus.getDefault().post(IAPFlowEvent(IAPFlowData.IAPAction.PURCHASE_FLOW_COMPLETE))
```

Comment:
Starts the flow from within a thread which further causes 'MainDashboardRefreshEvent' to crash on `WebViewDiscoverFragment`.

Reply:
fixed.